lib/data/wrapper.js

```diff
@@ -285,6 +288,173 @@ const data = {
+     * destination object
+     * dataStoreContext.bucketName: destination bucket name,
+     * dataStoreContext.owner: owner,
+     * dataStoreContext.namespace: request namespace,
+     * dataStoreContext.objectKey: destination object key name,
+     * @param {BackendInfo} destBackendInfo - Instance of BackendInfo:
+     * Represents the info necessary to evaluate which data backend to use
+     * on a data put call.
+     * @param {object} serverSideEncryption - serverSideEncryption
+     * @param {object} log - Werelogs request logger
+     * @param {function} cb - callback
+     * @returns {function} cb - callback
+     */
+    copyObject: (request, sourceLocationConstraintType,
+      sourceLocationConstraintName, storeMetadataParams, dataLocator,
+      dataStoreContext, destBackendInfo, serverSideEncryption, log, cb) => {
+        // NOTE: using copyObject only if copying object from one external
+        // backend to the same external backend
+        const locationTypeMatch = request.headers[locationHeader] ?
+        config.getLocationConstraintType(sourceLocationConstraintName) ===
+        config.getLocationConstraintType(request.headers[locationHeader])
+        : true;
+        if (locationTypeMatch &&
+        (sourceLocationConstraintType === 'aws_s3' ||
+        (sourceLocationConstraintType === 'azure' && config.isSameAzureAccount(
+        sourceLocationConstraintName, request.headers[locationHeader])))) {
+            const location = storeMetadataParams
+            .metaHeaders[locationHeader];
+            const objectGetInfo = dataLocator[0];
+            const externalSourceKey = objectGetInfo.key;
+            return client.copyObject(request, location,
+            externalSourceKey, sourceLocationConstraintName, log,
+            (error, objectRetrievalInfo) => {
+                if (error) {
+                    return cb(error);
+                }
+                const putResult = {
+                    key: objectRetrievalInfo.key,
+                    dataStoreName: objectRetrievalInfo.
+                        dataStoreName,
+                    dataStoreType: objectRetrievalInfo.
+                        dataStoreType,
+                    size: storeMetadataParams.size,
+                    dataStoreETag: objectGetInfo.dataStoreETag,
+                    start: objectGetInfo.start,
+                };
+                const putResultArr = [putResult];
+                return cb(null, putResultArr);
+            });
+        }
+
+        // dataLocator is an array.  need to get and put all parts
+        // For now, copy 1 part at a time. Could increase the second
+        // argument here to increase the number of parts
+        // copied at once.
+        return async.mapLimit(dataLocator, 1,
+            // eslint-disable-next-line prefer-arrow-callback
+            function copyPart(part, cb) {
+                if (part.dataStoreType === 'azure') {
```

Comment:
I see why you kept knowledge of azure here. @ploki will want to remove it but this is still a good first step to get the azure logic over to the wrapper and out of the api.